core/src/main/java/io/grpc/internal/Http2ClientStream.java

```diff
@@ -87,24 +86,34 @@ protected void transportHeadersReceived(Metadata headers) {
       transportError = transportError.augmentDescription(headers.toString());
       return;
     }
+
+    // Verify that the HTTP status is OK.
     Status httpStatus = statusFromHttpStatus(headers);
-    if (httpStatus == null) {
-      transportError = Status.INTERNAL.withDescription(
-          "received non-terminal headers with no :status");
-    } else if (!httpStatus.isOk()) {
-      transportError = httpStatus;
-    } else {
-      transportError = checkContentType(headers);
-    }
-    if (transportError != null) {
+    if (httpStatus == null || !httpStatus.isOk()) {
+      transportError = httpStatus != null ? httpStatus : Status.INTERNAL.withDescription(
+              "received non-terminal headers with no :status");
+
+      // TODO(nmittler): What to do if nothing else comes?
+      // The error is generated from the HTTP status.
       // Note we don't immediately report the transport error, instead we wait for more data on the
       // stream so we can accumulate more detail into the error before reporting it.
       transportError = transportError.augmentDescription("\n" + headers.toString());
       errorCharset = extractCharset(headers);
-    } else {
-      stripTransportDetails(headers);
-      inboundHeadersReceived(headers);
+      return;
     }
+
+    // Verify that the content type is appropriate for gRPC.
+    transportError = checkContentType(headers);
+    if (transportError != null) {
+      // The request contained an invalid Content-Type, cancel the stream now.
```

Comment:
The thought here is that if we're receiving a successful response with an incorrect content-type, we have no idea how much data might follow.  Probably best to just terminate the stream now.

This is different than if we receive an HTTP response with an error. In that case, it's likely that whatever comes afterward will be error/debug information that might be useful.